x-pack/plugins/upgrade_assistant/server/lib/reindexing/reindex_actions.ts

```diff
@@ -0,0 +1,372 @@
+   * @param reindexOp
+   * @param attrs
+   */
+  updateReindexOp(
+    reindexOp: ReindexSavedObject,
+    attrs?: Partial<ReindexOperation>
+  ): Promise<ReindexSavedObject>;
+
+  /**
+   * Runs a callback function while locking the reindex operation. Guaranteed to unlock the reindex operation when complete.
+   * @param func A function to run with the locked ML lock document. Must return a promise that resolves
+   * to the updated ReindexSavedObject.
+   */
+  runWhileLocked(
+    reindexOp: ReindexSavedObject,
+    func: (reindexOp: ReindexSavedObject) => Promise<ReindexSavedObject>
+  ): Promise<ReindexSavedObject>;
+
+  /**
+   * Finds the reindex operation saved object for the given index.
+   * @param indexName
+   */
+  findReindexOperations(indexName: string): Promise<FindResponse<ReindexOperation>>;
+
+  /**
+   * Returns an array of all reindex operations that have a status.
+   */
+  findAllByStatus(status: ReindexStatus): Promise<ReindexSavedObject[]>;
+
+  /**
+   * Returns array of field paths to boolean fields in the index's mapping.
+   * @param indexName
+   */
+  getBooleanFieldPaths(indexName: string): Promise<string[][]>;
+
+  /**
+   * Retrieve index settings (in flat, dot-notation style) and mappings.
+   * @param indexName
+   */
+  getFlatSettings(indexName: string): Promise<FlatSettings | null>;
+
+  /**
+   * Tries to undo any changes that were made in the event of a failure.
+   * TODO: move back to reindex service?
+   * @param indexName
+   */
+  cleanupChanges(indexName: string): Promise<void>;
+
+  /**
+   * Returns whether or not an indexName is a special ML index that requires suspending ML jobs.
+   * @param indexName
+   */
+  isMlIndex(indexName: string): boolean;
+
+  // ----- Below are only for ML indices
+
+  /**
+   * Atomically increments the number of reindex operations running for ML jobs.
+   */
+  incrementMlReindexes(): Promise<void>;
+
+  /**
+   * Atomically decrements the number of reindex operations running for ML jobs.
+   */
+  decrementMlReindexes(): Promise<void>;
+
+  /**
+   * Runs a callback function while locking the ML count.
+   * @param func A function to run with the locked ML lock document. Must return a promise that resolves
+   * to the updated ReindexSavedObject.
+   */
+  runWhileMlLocked(
+    func: (mlLockDoc: ReindexSavedObject) => Promise<ReindexSavedObject>
+  ): Promise<void>;
+
+  /**
+   * Exposed only for testing, DO NOT USE.
+   */
+  _fetchAndLockMlDoc(): Promise<ReindexSavedObject>;
+}
+
+export const reindexActionsFactory = (
+  client: SavedObjectsClient,
+  callCluster: CallCluster
+): ReindexActions => {
+  // ----- Internal functions
+  /**
+   * Generates a new index name for the new index. Iterates until it finds an index
+   * that doesn't already exist.
+   * @param indexName
+   */
+  const getNewIndexName = async (indexName: string, attempts = 100) => {
+    for (const i of range(0, attempts)) {
+      const newIndexName = `${indexName}-reindex-${i}`;
```

Comment:
Should we remove `-reindex-X` from the indexName and start at X? In 7.0 users could end up with `foo-reindex-0-reindex-0`

Reply:
In 5.6 we appended `-reindexed-v5`. I wonder if we should also remove it from the indexName. 

Additionally, thoughts on `reindexed` instead of `reindex` and also starting at `1` instead of `0`.

Reply:
Yeah let's just change to the same pattern used in 5.6 and replace v5 with v6 if it's already in the index name.